src/parser/mod.rs

```diff
@@ -11096,6 +11096,19 @@ impl<'a> Parser<'a> {
                 arg,
                 operator: FunctionArgOperator::Assignment,
             })
+        } else if dialect_of!(self is MsSqlDialect) {
+            // FUNC(<expr> : <expr>)
+            let name = self.parse_wildcard_expr()?;
+            if self.consume_token(&Token::Colon) {
+                let arg = self.parse_expr()?.into();
+                Ok(FunctionArg::Named {
+                    name,
+                    arg,
+                    operator: FunctionArgOperator::Colon,
+                })
+            } else {
+                Ok(FunctionArg::Unnamed(name.into()))
+            }
```

Comment:
```suggestion
        } else if self.dialect.supports_json_object_functions() && self.peek_nth_token(1) == Token::Colon {
            let name = self.parse_wildcard_expr(false)?;
            let _ = self.expect_token(&Token::Colon)?;
            let arg = self.parse_expr()?.into();
            Ok(FunctionArg::Named {
                    name,
                    arg,
                    operator: FunctionArgOperator::Colon,
                })
            }
```
Something like this so that we only enter the branch when we know that there is a valid expr to parse? We can use a method on the dialect for this feature instead of the `dialect_of!` macro (the latter is essentially deprecated)

Reply:
`self.peek_nth_token(1)` cannot capture the case where the name is an expression.

Reply:
Ah that right! that makes sense, though can we collapse the last else if /else so that the unnamed default is constructed once?
```rust
} else  {
  let name = ...;
  if dialect_of!(self is MsSqlDialect) {
    return Ok(FunctionArg::Named{...})
  }
  Ok(FunctionArg::Unnamed(name.into()))
}
```